File: kun-metadata/src/main/java/com/miotech/kun/metadata/client/JDBCClient.java
package com.miotech.kun.metadata.client;

import com.google.common.collect.Maps;
import com.miotech.kun.metadata.constant.DatabaseType;
import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.sql.DataSource;
import java.sql.*;
import java.util.Map;

/**
 * Universal JDBC connection tool, which adapts to Hive and Postgres
 * @author zhangxinwei
 */
public class JDBCClient {
    private static Logger logger = LoggerFactory.getLogger(JDBCClient.class);
    private static final Map<String, DataSource> dataSourceCache = Maps.newConcurrentMap();

    public static Connection getConnection(DatabaseType dbType, String url, String user, String password)
            throws ClassNotFoundException, SQLException {
        if (dbType == null) {
            throw new RuntimeException("dbType must not be null");
        }

        String driverName = selectSpecificDriver(dbType);
        Class.forName(driverName);
        return DriverManager.getConnection(url, user, password);
    }

    public static DataSource getDataSource(String url, String username, String password, DatabaseType dbType) {
        String key = url + username + password;
        if (dataSourceCache.containsKey(key)) {
            return dataSourceCache.get(key);
        }
        HikariConfig config = new HikariConfig();
        config.setJdbcUrl(url);
        config.setUsername(username);
        config.setPassword(password);
        config.setDriverClassName(selectSpecificDriver(dbType));
        HikariDataSource hikariDataSource = new HikariDataSource(config);

        dataSourceCache.put(key, hikariDataSource);
        return hikariDataSource;
    }

    private static String selectSpecificDriver(DatabaseType dbType) {
        switch (dbType) {
            case HIVE:
                return "org.apache.hive.jdbc.HiveDriver";
            case MYSQL:
                return "com.mysql.jdbc.Driver";
            case POSTGRES:
                return "org.postgresql.Driver";
            case PRESTO:
                return "io.prestosql.jdbc.PrestoDriver";
            case ATHENA:
                return "com.simba.athena.jdbc.Driver";
            default:
                throw new RuntimeException("invalid dbType: " + dbType.getName());
        }
    }

    public static void close(Connection connection, Statement statement, ResultSet resultSet) {
        try {
            if (resultSet != null) {
                resultSet.close();
            }
            if (statement != null) {
                statement.close();
            }
            if (connection != null) {
                connection.close();
            }
        } catch (SQLException sqlException) {
            logger.error("JDBCClient release db resources exception", sqlException);
        }
    }

}

File: kun-workflow/kun-workflow-operators/spark-etl-operator/src/main/scala/com/miotech/spark/etl/common/JdbcClient.scala
package com.miotech.spark.etl.common

import java.sql.{Connection, DriverManager, SQLException}

import org.slf4j.LoggerFactory

object JdbcClient {

  def apply(dbTypeName: String,
            url: String,
            username: String,
            password: String
           ) = {
    val dbType = dbTypeName match {
      case "postgresql" => POSTGRESQL
      case "mysql" => MYSQL
      case "sqlserver" => SQLSERVER
    }
    new JdbcClient(dbType, url, username, password)
  }
}

class JdbcClient( dbType: DBType,
                  url: String,
                  username: String,
                  password: String
                ) {

  private val logger = LoggerFactory.getLogger(getClass)

  private var conn: Connection = _

  private val jdbcDriverMap = Map(
    POSTGRESQL.name -> "org.postgresql.Driver",
    MYSQL.name -> "com.mysql.cj.jdbc.Driver",
    SQLSERVER.name -> "com.microsoft.sqlserver.jdbc.SQLServerDriver"
  )

  val driver: String = jdbcDriverMap(dbType.name)

  def getConnection(): Connection = {
    if (conn == null || conn.isClosed) {
      Class.forName(driver)
      DriverManager.getConnection(url, username, password)
    } else conn
  }

  def execute(sql:String, transactional:Boolean = true): Unit = {
    logger.debug("sql:", sql)
    val conn = getConnection()

    val start = System.currentTimeMillis()
    try {
      if (transactional) {
        conn.setAutoCommit(false)
        conn.createStatement().execute(sql)
        conn.commit()
      } else {
        conn.createStatement().execute(sql)
      }
    } catch {
      case e: SQLException =>
        logger.error("{}", e)
        throw e
    }
    val end = System.currentTimeMillis()
    logger.debug(s"takes: ${end - start} ms")
  }
}
